# 08 EDA — Geolocation

Read-only inspection of `olist_geolocation_dataset.csv`. No cleaning, charts, or feature engineering.

## 1. Import libraries

Import pandas for tabular inspection.

In [1]:
import pandas as pd

**Observation:** pandas is ready.

## 2. Load CSV

Load only this CSV into memory without changing the source.

In [2]:
file_path = '../data/raw/olist/olist_geolocation_dataset.csv'
df = pd.read_csv(file_path)

**Observation:** The table is loaded; the raw CSV is unchanged.

## 3. Shape

Check table size.

In [3]:
df.shape

(1000163, 5)

**Observation:** This reports the table's row and column count.

## 4. First 5 rows

Preview the raw structure.

In [4]:
df.head()

   geolocation_zip_code_prefix  ...  geolocation_state
0                         1037  ...                 SP
1                         1046  ...                 SP
2                         1046  ...                 SP
3                         1041  ...                 SP
4                         1035  ...                 SP

[5 rows x 5 columns]

**Observation:** The preview confirms the row grain and field layout.

## 5. Column names

List available fields.

In [5]:
df.columns.tolist()

['geolocation_zip_code_prefix', 'geolocation_lat', 'geolocation_lng', 'geolocation_city', 'geolocation_state']

**Observation:** These are the fields available for this table's analysis.

## 6. Data types

Review pandas data types.

In [6]:
df.dtypes

geolocation_zip_code_prefix      int64
geolocation_lat                float64
geolocation_lng                float64
geolocation_city                   str
geolocation_state                  str
dtype: object

**Observation:** Types show which fields are text, numeric, or converted datetime values.

## 7. Missing values

Count nulls by field.

In [7]:
df.isna().sum()

geolocation_zip_code_prefix    0
geolocation_lat                0
geolocation_lng                0
geolocation_city               0
geolocation_state              0
dtype: int64

**Observation:** Missing counts identify incomplete fields without changing them.

## 8. Missing percentage

Measure missingness relative to table size.

In [8]:
(df.isna().mean().mul(100).round(2)).sort_values(ascending=False)

geolocation_zip_code_prefix    0.0
geolocation_lat                0.0
geolocation_lng                0.0
geolocation_city               0.0
geolocation_state              0.0
dtype: float64

**Observation:** Percentages show the relative scale of incomplete fields.

## 9. Duplicate rows

Count complete row duplicates.

In [9]:
df.duplicated().sum()

np.int64(261831)

**Observation:** This is the number of repeated full rows; none are removed.

## 10. Duplicate primary keys

Validate the documented candidate key.

In [10]:
candidate_key = ['geolocation_zip_code_prefix']
df.duplicated(subset=candidate_key, keep=False).sum()

np.int64(999120)

**Observation:** Geolocation has no verified natural primary key; this shows how many rows share a ZIP prefix.

## 11. Unique values

Inspect meaningful category or entity counts.

In [11]:
pd.Series({'zip_prefixes': df['geolocation_zip_code_prefix'].nunique(), 'states': df['geolocation_state'].nunique(), 'cities': df['geolocation_city'].nunique()})

zip_prefixes    19015
states             27
cities           8011
dtype: int64

**Observation:** This gives a concise view of important distinct values.

## 12. Basic describe()

Generate a basic summary for all fields.

In [12]:
df.describe(include='all')

        geolocation_zip_code_prefix  ...  geolocation_state
count                  1.000163e+06  ...            1000163
unique                          NaN  ...                 27
top                             NaN  ...                 SP
freq                            NaN  ...             404268
mean                   3.657417e+04  ...                NaN
std                    3.054934e+04  ...                NaN
min                    1.001000e+03  ...                NaN
25%                    1.107500e+04  ...                NaN
50%                    2.653000e+04  ...                NaN
75%                    6.350400e+04  ...                NaN
max                    9.999000e+04  ...                NaN

[11 rows x 5 columns]

**Observation:** The summary describes completeness, distinctness, common values, and numeric ranges where applicable.

## 13. Table-specific observations

Geolocation contains coordinate observations for Brazilian ZIP prefixes. Multiple observations can represent the same prefix.

## 14. Data quality observations

There are 261,831 exact duplicate rows and no valid raw natural primary key. ZIP prefix has only 19,015 distinct values across 1,000,163 rows, so direct joins can multiply business records.